lib/goal/container/k8s.ts

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    GitCommandGitProject,
+    GitProject,
+    HandlerContext,
+    logger,
+} from "@atomist/automation-client";
+import { sleep } from "@atomist/automation-client/lib/internal/util/poll";
+import {
+    DefaultGoalNameGenerator,
+    doWithProject,
+    ExecuteGoal,
+    GoalScheduler,
+    ImplementationRegistration,
+    ProgressLog,
+    RepoContext,
+    SdmGoalEvent,
+    SdmGoalState,
+    ServiceRegistrationGoalDataKey,
+} from "@atomist/sdm";
+import * as k8s from "@kubernetes/client-node";
+import * as stringify from "json-stringify-safe";
+import * as _ from "lodash";
+import * as os from "os";
+import * as request from "request";
+import { Writable } from "stream";
+import {
+    DeepPartial,
+    Merge,
+} from "ts-essentials";
+import { loadKubeConfig } from "../../pack/k8s/config";
+import {
+    k8sJobEnv,
+    KubernetesGoalScheduler,
+    readNamespace,
+} from "../../pack/k8s/KubernetesGoalScheduler";
+import {
+    K8sServiceRegistrationType,
+    K8sServiceSpec,
+} from "../../pack/k8s/service";
+import { toArray } from "../../util/misc/array";
+import {
+    Container,
+    ContainerProjectHome,
+    ContainerRegistration,
+    ContainerScheduler,
+    GoalContainer,
+    GoalContainerVolume,
+} from "./container";
+import {
+    containerEnvVars,
+    copyProject,
+    loglog,
+} from "./util";
+
+/**
+ * Specification of containers and volumes for a container goal.
+ */
+export interface K8sGoalContainerSpec {
+    /**
+     * Containers to run for this goal.  The goal result is based on
+     * the exit status of the first element of the `containers` array.
+     * The other containers are considered "sidecar" containers
+     * provided functionality that the main container needs to
+     * function.  The working directory of the first container is set
+     * to [[ContainerProjectHome]], which contains the project upon
+     * which the goal should operate.
+     */
+    containers: Array<Merge<DeepPartial<k8s.V1Container>, GoalContainer>>;
```

Comment:
Here the same question: why did you opt to separate `containers` and `callback`?

Reply:
Same reasons as above.